src/box/txn.c

```diff
@@ -787,6 +788,7 @@ txn_complete_success(struct txn *txn)
 	if (txn->engine != NULL)
 		engine_commit(txn->engine, txn);
 	if (txn_has_flag(txn, TXN_HAS_TRIGGERS)) {
+		assert(in_txn() != NULL);
```

Comment:
Shouldn't it be `assert(in_txn() == txn)`?

> Currently any transaction on synchronous space fails to complete with
the ER_CURSOR_NO_TRANSACTION error, when on_rollback/on_commit triggers
are set. This is caused due to the fact, that some rollback/commit
triggers require in_txn fiber variable to be set but it's not done
when a transaction is completed from the limbo.

Please elaborate. What triggers? How exactly does this error occur?

AFAICS a commit/rollback trigger callback is passed the `txn` as an argument so why don't you patch the triggers that use `in_txn()` so that they use the passed `txn` instead?

Reply:
> > Please elaborate. What triggers? How exactly does this error occur?
> 
> Everything is already described in the issue. It's `lbox_txn_pairs` (and many other callbacks) function, which opens an iterator over txn's satements. It has nothing but `txn_id` on the Lua stack. The only way it can get `txn` from is fiber. I see no reason to rewrite all triggers, moreover, I'm not sure, we can push txn to stack at all.

OK. Please describe it in the commit message as well.

> > Shouldn't it be assert(in_txn() == txn)?
> 
> It should not, we must be sure, that txn is set, when triggers are run. A lot of functions in `txn.c` set `in_txn` just to be sure, that `in_txn() != NULL`.

How can `in_txn()` be different from `txn` here? If so, it looks like a bug because this would mean that `lbox_txn_pairs()` would run on a different transaction. Also, this assertion should probably be unconditional (i.e. run even if `TXN_HASH_TRIGGERS` is unset).

Reply:
> How can in_txn() be different from txn here?

I misread your message and thought you meant `assert(in_txn() == NULL)`. I agree, that we should add `assert(in_txn() == txn)`.

> Please describe it in the commit message as well.

Extended description in commit message